File: server/routers/studio.py
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

import crud_ops as crud
import schemas
from dependencies import get_current_user_id, get_db

router = APIRouter(prefix="/api/studio", tags=["studio"])


@router.get("/bootstrap", response_model=schemas.StudioBootstrapResponse)
def read_studio_bootstrap(
    limit: int = 24,
    db: Session = Depends(get_db),
    owner_id: str = Depends(get_current_user_id),
):
    return crud.get_studio_bootstrap(db, owner_id, limit=limit)


@router.get("/scripts", response_model=schemas.StudioScriptsResponse)
def read_studio_scripts(
    limit: int = 24,
    offset: int = 0,
    status: str = "all",
    q: str = "",
    sort: str = "updated_desc",
    includeCounts: bool = True,
    db: Session = Depends(get_db),
    owner_id: str = Depends(get_current_user_id),
):
    return crud.list_studio_scripts(
        db,
        owner_id,
        limit=limit,
        offset=offset,
        status_filter=status,
        search=q,
        sort=sort,
        include_counts=includeCounts,
    )


@router.get("/scripts/{script_id}/publish-context", response_model=schemas.StudioScriptSummary)
def read_studio_publish_context(
    script_id: str,
    db: Session = Depends(get_db),
    owner_id: str = Depends(get_current_user_id),
):
    item = crud.get_studio_publish_context(db, owner_id, script_id)
    if not item:
        raise HTTPException(status_code=404, detail="Script not found")
    return item

File: src/lib/api/studio.ts
import { fetchApi } from "./client";
import type { BaseScriptApi, StudioBootstrapResponse, StudioScriptsResponse } from "../../types/api";

export interface StudioScriptsQuery {
  limit?: number;
  offset?: number;
  status?: string;
  q?: string;
  sort?: string;
  includeCounts?: boolean;
}

const toQueryString = (query: StudioScriptsQuery): string => {
  const params = new URLSearchParams();
  if (typeof query.limit === "number") params.set("limit", String(query.limit));
  if (typeof query.offset === "number") params.set("offset", String(query.offset));
  if (query.status && query.status !== "all") params.set("status", query.status);
  if (query.q?.trim()) params.set("q", query.q.trim());
  if (query.sort) params.set("sort", query.sort);
  if (typeof query.includeCounts === "boolean") params.set("includeCounts", query.includeCounts ? "true" : "false");
  const qs = params.toString();
  return qs ? `?${qs}` : "";
};

export const getStudioBootstrap = async (limit = 24): Promise<StudioBootstrapResponse> => {
  return fetchApi<StudioBootstrapResponse>(`/studio/bootstrap?limit=${encodeURIComponent(String(limit))}`);
};

export const getStudioScriptsPage = async (query: StudioScriptsQuery = {}): Promise<StudioScriptsResponse> => {
  return fetchApi<StudioScriptsResponse>(`/studio/scripts${toQueryString(query)}`);
};

export const getStudioPublishContext = async (scriptId: string): Promise<BaseScriptApi> => {
  return fetchApi<BaseScriptApi>(`/studio/scripts/${encodeURIComponent(scriptId)}/publish-context`);
};
